man/curve-keygen.1
.\" Automatically generated by Pod::Man 2.25 (Pod::Simple 3.16)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "CURVE-KEYGEN 1"
.TH CURVE-KEYGEN 1 "2013-10-07" "CURVE-KEYGEN 0.0.1" "USER CONTRIBUTED DOCUMENTATION"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
curve\-keygen \- a utility to generate CURVE25519 keypairs.
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\& Usage: curve\-genkey [options]
\&
\& Options:
\&  \-f <keyfile>    Filename for the key, default: \*(Aqid_curve25519\*(Aq.
\&  \-n <name>       Your name
\&  \-m <mail>       Your email address
\&  \-o <org>        Your organization
\&  \-y              Read private key file and print public key, requires \-s
\&  \-s <secretfile> File containing a secret key
\&  \-c              Don\*(Aqt encrypt the secret key (CZMQ default)
\&  \-h              print this help message
\&  \-v              print program version
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBcurve-genkey\fR generates a \s-1CURVE25519\s0 keypair. It stores the public
key in \s-1CZMQ\s0's zcert format and the private key in binary form. If
the option \fB\-c\fR have been supplied or if no passphrase provided, the
private key will be stored in zcert format as well.
.SH "LIBRARY"
.IX Header "LIBRARY"
A library \fBlibcsk\fR is available which can be used from client programs
to access encrypted private keys. Example usage:
.PP
.Vb 2
\& #include <csk.h>
\& #include <czmq.h>
\& 
\& int main() {
\&   size_t rawsize;
\&   char secretfile[] = "id_curve25519_secret";
\&   char passphrase[] = "my password";
\&   unsigned char *raw = csk_raw_load(secretfile, &rawsize);
\&   zcert_t *cert = csk_zcert_decrypt(raw, passphrase, rawsize);
\&   if(cert != NULL) {
\&     // do something with cert
\&   }
\&   return 0;
\& }
.Ve
.PP
Compile the client program with \fB\-lcsk\fR ld flag.
.SH "FILES"
.IX Header "FILES"
\&\fBcurve-keygen\fR creates by default files in the current directory. If
not further specified these are:
.IP "\fBid_curve25519_secret\fR" 4
.IX Item "id_curve25519_secret"
The encrypted private key in binary form. If the private key
is saved without encryption this file will be in \s-1CZMQ\s0 zcert format.
.IP "\fBid_curve25519\fR" 4
.IX Item "id_curve25519"
The public key in \s-1CZMQ\s0 zcert format.
.SH "BUGS"
.IX Header "BUGS"
Currently \fBcurve-keygen\fR stores the encrypted private key as
raw binary stream to disk. This is not portable. A future version
will save the key in \s-1IETF\s0 format or \- once \s-1CZMQ\s0 provides support for it \- 
in zcert-secret-format (however that might be named then).
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2013 by T.Linden <tom \s-1AT\s0 cpan \s-1DOT\s0 org>
.SH "AUTHORS"
.IX Header "AUTHORS"
\&\fIT.Linden <tom \s-1AT\s0 cpan \s-1DOT\s0 org\fR>
.SH "LICENSE"
.IX Header "LICENSE"
Licensed under the  \s-1GNU\s0 \s-1GENERAL\s0 \s-1PUBLIC\s0 \s-1LICENSE\s0 version 3.


ChangeLog
0.01   initial version


src/curve-keygen.c
/*
    curve-genkey
    
    Certificate generator for ZeroMQ CURVE security. Produces two files:
    
    * id_curve25519        - public certificate
    * id_curve25519.secret - secret certificate
    
    Copyright (C) 2013 iMatix Corporation 
    Copyright (c) 2013 T.Linden

    Licensed under MIT/X11.
    
    Permission is hereby granted, free of charge, to any person obtaining 
    a copy of this software and associated documentation files (the 
    "Software"), to deal in the Software without restriction, including 
    without limitation the rights to use, copy, modify, merge, publish, 
    distribute, sublicense, and/or sell copies of the Software, and to 
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be 
    included in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS 
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF 
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. 
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY 
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, 
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE 
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

#include "curve-keygen.h"
#include "getpass.h"
#include "version.h"
  



void usage() {
  fprintf(stderr, "Usage: curve-genkey [options]\n");
  fprintf(stderr,  "Options:\n");
  fprintf(stderr,  "  -f <keyfile>    Filename for the key, default: 'id_curve25519'.\n");
  fprintf(stderr,  "  -n <name>       Your name\n");
  fprintf(stderr,  "  -m <mail>       Your email address\n");
  fprintf(stderr,  "  -o <org>        Your organization\n");
  fprintf(stderr,  "  -y              Read private key file and print public key, requires -s\n");
  fprintf(stderr,  "  -s <secretfile> File containing a secret key\n");
  fprintf(stderr,  "  -c              Don't encrypt the secret key (CZMQ default)\n");
  fprintf(stderr,  "  -h              print this help message\n");
  fprintf(stderr,  "  -v              print program version\n");
  exit(EXIT_FAILURE);
}

void version() {
  fprintf(stderr, "curve-keygen version %d.%d.%d\n", CSK_VERSION_MAJOR, CSK_VERSION_MINOR, CSK_VERSION_PATCH);
  exit(0);
}

int main (int argc, char **argv)  {
  int opt;
  char *filename = "id_curve25519";
  char *secretfile = "";
  char *name     = "";
  char *mail     = "";
  char *org      = "";
  int export     = 0;
  int unsecure   = 0;
  int filenamegiven = 0;

  while (1) {
    opt = getopt(argc, argv, "vf:n:m:o:hys:");

    if(opt == -1) {
      break;
    }

    switch (opt)  {
      case 'y':
	export = 1;
	break;
      case 'c':
	unsecure = 1;
	break;
      case 'f':
	filename = optarg;
	filenamegiven = 1;
	break;
      case 's':
	secretfile = optarg;
	break;	
      case 'n':
	name = optarg;
	break;
      case 'm':
	mail = optarg;
	break;
      case 'o':
	org = optarg;
	break;
      case 'h':
	usage();
      case 'v':
        version();
      default:
	usage();
    }
  }

  if(export) {
    // export the public key from a private one
    if(secretfile[0] != '\0' && filename[0] != '0' && filenamegiven) {
      if(! zfile_exists(secretfile)) {
	fprintf(stderr, "Private key file %s doesn't exist!\n", secretfile);
	return -1;
      }
      else {
	int ok = 0;
	size_t rawsize;
	unsigned char *privraw = csk_raw_load(secretfile, &rawsize);

	if(privraw != NULL) {
	  char *passphrase;

	  if(strncmp((char *)privraw, "# ", 2) == 0) {
	    // not encrypted
	    zcert_t *priv = zcert_load(secretfile);
	    zcert_save_public(priv, filename);
	    free(priv);
	  }
	  else {
	    passphrase = csk_get_passphrase("Enter passphrase");
	    if(passphrase == NULL || strlen(passphrase) <= 0) {
	      fprintf(stderr, "Sorry, you need to enter a passphrase in order to decrypt the secret key!\n");
	      ok = -1;
	    }
	    else {
	      zcert_t *privclear = csk_zcert_decrypt(privraw, passphrase, rawsize);
	      if (privclear != NULL) {
		zcert_save_public(privclear, filename);
		printf ("Public key from Curve25519 certificate %s exported to %s\n", secretfile, filename);
		free(privclear);
	      }
	      else {
		ok = -1;
	      }
	    }
	  }

	  free(privraw);
	  free(passphrase);
	}
	else {
	  ok = -1;
	}

	return ok;
      }
    }
    else {
      fprintf(stderr, "Parameter -s <secretkeyfile> (input) and -f <publickeyfile> (output) required for -y!\n");
      return -1;
    }
  }

  /* make sure we don't destroy something */
  if(zfile_exists(filename)) {
    printf("Key file %s already exists. Overwrite [Ny]? ", filename);
    char *c = csk_get_stdin();
    if(strncmp(c, "y", 80) != 0) {
      printf("abort\n");
      return 0;
    }
  }

  printf ("Creating new Curve25519 certificate %s\n", filename);

  zcert_t *cert = zcert_new ();

  if(name[0] == '\0') {
    if(s_get_meta (cert, "Enter your full name:", "name"))
      return -1;
  }
  else {
    zcert_set_meta(cert, "name", name);
  }

  if(mail[0] == '\0') {
    if(s_get_meta (cert, "Enter your email address:", "email"))
      return -1;
  }
  else {
    zcert_set_meta(cert, "email", mail);
  }

  if(org[0] == '\0'){
    if(s_get_meta (cert, "Enter your organization:", "org"))
      return -1;
  }
  else {
    zcert_set_meta(cert, "org", org);
  }
        
  char *timestr = zclock_timestr ();
  zcert_set_meta (cert, "created-by", "curve-genkey");
  zcert_set_meta (cert, "date-created", timestr);
  free (timestr);
  char *sec = "Unsecure";

  if (unsecure) {
    /* create an original CZMQ cert with unencrypted secret key */
    zcert_save (cert, filename);
  }
  else {
    char *pass1 = csk_get_passphrase("Enter a passphrase, leave empty for unprotected key");
    if(pass1 != NULL && strlen(pass1) > 0) {
      char *pass2 = csk_get_passphrase("                           Enter a passphrase again");
      if(strncmp(pass1, pass2, 80) == 0) {
	size_t clen;
	unsigned char *encrypted = csk_zcert_encrypt(cert, pass1, &clen);
	if(encrypted != NULL) {
	  char *sfile = ucmalloc(strlen(filename) + 8);
	  strncpy(sfile, filename, strlen(filename) + 7);
	  strcat(sfile, "_secret");
	  if((csk_raw_save (encrypted, sfile, clen)) == 0) {
	    zcert_save_public(cert, filename);
	    sec = "Encrypted";
	  }
	  else {
	    return -1;
	  }
	}
	else {
	  return -1;
	}
      }
      else {
	fprintf(stderr, "Passphrases were not identical, please repeat!\n");
	return -1;
      }
    }
    else {
      /* user just pressed enter, so make default CZMQ unencrypted cert */
      zcert_save (cert, filename);
    }
  }

  printf ("%s Curve25519 certificate created in %s and %s_secret\n", sec, filename, filename);
  
  zcert_destroy (&cert);

  return 0;
}
